### PR TITLE
sql: user/role refactor to match postgres

### PR DESCRIPTION
**Pull request by RichardJCai**
Refactoring Users and Roles to be the same. (Same as PG)

Release note (sql-feature): CREATE ROLE / CREATE USER are now the same except:
CREATE ROLE would have login disabled by default and CREATE USER has login enabled by default.
(login privilege is not added in this PR). Similarly ALTER/DROP ROLE and USER are the same.

CREATE/ALTER/DROP ROLE syntax are still disabled for users despite CREATE USER being an alias for them.
Similarly DROP / ALTER are the same for roles and users.
Done to match PG behaviour.

-------
Discussion:

Questions for product:
Need to determine how to handle enterprise / non enterprise features.
Would it still make sense to disable roles for non enterprise?
One option is to disable user level privileges (CREATEROLE/LOGIN) and GRANTING users/roles to users/roles (This is still currently disabled).

Is it necessary to disable creation of "roles" (the syntax CREATE ROLE) if GRANTING users/roles is disabled?

CREATE USER / ROLE is effectively the same thing.
(In postgres - CREATE USER is an alias for CREATE ROLE)

What should SHOW USERS / SHOW ROLES do? (Equivalent?)

Edit:
After some discussion, we decided CREATE/ALTER/DROP ROLE will still be disabled for non-enterprise to avoid confusion (still cannot GRANT) "roles" don't really exist.
SHOW USERS will show roles/users that can login. 
SHOW ROLES will show all roles and users. 

- Will handle the SHOW logic in the LOGIN option PR since it depends on the LOGIN flag.

### CODE REVIEW

**Review by solongordon — CHANGES_REQUESTED** [review]
I took a first pass. Looks like this is headed in a good direction. Main points I want to look at:
- Should we keep CREATE USER and CREATE ROLE separate at the syntax level?
- Is it possible to reuse `KVOption` instead of adding OptionWithValue?

> Reviewed 33 of 33 files at r1, 34 of 34 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-:-M-pOpxs2KODZL38qntM:beh0o5a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[docs/generated/settings/settings.html, line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pBoepCPb4eLqynASB:-M-pBoeq64iXt9ZZSQ5G:b-fpvqt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/docs/generated/settings/settings.html#L0)):*
Looks like this file went away for some reason?
___
*[pkg/ccl/logictestccl/testdata/logic_test/role, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pC2Ub0zl1-xCfaRU0:-M-pC2Ub0zl1-xCfaRU1:bwc9mat) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43f506594e7561475cad4bf712c3a3329f0de35/pkg/ccl/logictestccl/testdata/logic_test/role#L53)):*
> ```
> 
> statement error pq: user myrole does not exist
> DROP USER myrole
> ```

Maybe we should still have a test for this if we intend for roles to be droppable with DROP USER?
___
*[pkg/ccl/roleccl/role.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGK9_2AAVxWO34Yim:-M-pGK9_2AAVxWO34Yin:bg7wofb) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/ccl/roleccl/role.go#L116)):*
> ```Go
> 		// TODO(mberhault): just like GRANT/REVOKE privileges, we fetch the list of all users.
> 		// This is wasteful when we have a LOT of users compared to the number of users being operated on.
> 		users, err := p.GetAllUsersAndRoles(ctx)
> ```

Should we just change the function name to `GetAllRoles`? And does it still need to return whether each is a user or a role?
___
*[pkg/ccl/roleccl/role.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGHL84IAESvzIZ4yA:-M-pGHL84IAESvzIZ4yB:b-65ult5) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/ccl/roleccl/role.go#L125)):*
> ```Go
> 
> 		// Check roles: these have to be roles.
> 		// Updated to match pg, can be roles or users.
> ```

You can just update the comment.
___
*[pkg/sql/backfill.go, line 834 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pHJRiADdertfMFxrV:-M-pHJRiADdertfMFxrW:b-2m30zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/backfill.go#L834)):*
> ```Go
> 		// since they just add their keys blindly. Running a Scan of the target
> 		// spans at the time the SSTs' keys will be written will calcify history up
> 		// to then since the scan will resolveUsername intents and populate tscache to keep
> ```

Bad rename? Many more examples of this in other files too. I won't comment on them all.
___
*[pkg/sql/database.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pILDX0XLi-sb34OA6:-M-pILDX0XLi-sb34OA7:b-e55vtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/database.go#L237)):*
> ```Go
> 
> // getCachedDatabaseID returns the ID of a database given its name
> // from the cache. This method never goes to the store to resolveUsername
> ```

Bad rename
___
*[pkg/sql/parser/help_test.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJ-iZ8BKtivKaqiU3:-M-pJ-iZ8BKtivKaqiU4:bufzzvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/help_test.go#L69)):*
> ```Go
> 		{`ALTER USER foo WITH PASSWORD ??`, `ALTER USER`},
> 
> 		{`ALTER ROLE bleh ?? WITH NOCREATEROLE`, `ALTER USER`},
> ```

This looks weird. If anything I would expect to show the ALTER ROLE help for both ALTER ROLE and ALTER USER. Though it might be nicer to keep them separate.
___
*[pkg/sql/parser/parse_test.go, line 1919 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJDih8_-5nKa0unGe:-M-pJDih8_-5nKa0unGf:b-eabt7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/parse_test.go#L1919)):*
> ```Go
> 			`CREATE USER IF NOT EXISTS 'foo'`},
> 		{`CREATE USER foo PASSWORD bar`,
> 			`CREATE USER 'foo' PASSWORD 'bar'`},
> ```

I still think it's fine if we just insert the WITH when it's not specified. *shrug*
___
*[pkg/sql/parser/sql.y, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pLKwi2RikTWY9qok_:-M-pLKwi2RikTWY9qoka:b-8doen) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L346)):*
> ```Yacc
>     return u.val.(tree.OptionsWithValues)
> }
> func (u *sqlSymUnion) isRole() bool {
> ```

I don't think you need to add this. You can just use the existing `bool()` method.
___
*[pkg/sql/parser/sql.y, line 5072 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJX-p2aB7OJOBi1DY:-M-pJX-p2aB7OJOBi1DZ:b-nya8i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5072)):*
> ```Yacc
> // %Text: CREATE USER [IF NOT EXISTS] <name> [WITH] <OPTIONS...>
> // %SeeAlso: ALTER ROLE, DROP ROLE, SHOW ROLES, ALTER USER, DROP USER, SHOW USER
> create_user_or_role_stmt:
> ```

It's unclear to me whether we should be merging these statements at the syntax level. CREATE USER is still slightly different from CREATE ROLE and we probably want to show different help text for the two. Do you think perhaps we should keep them separate here even though the underlying implementation is going to be the same?
___
*[pkg/sql/parser/sql.y, line 5081 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pMgN04Oxm1QWpNKg-:-M-pMgN04Oxm1QWpNKg0:b-jt4bkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5081)):*
> ```Yacc
>     $$.val = &tree.CreateUserOrRole{Name: $6.expr(), IsRole: $2.isRole(), IfNotExists: true}
>   }
> | CREATE role_or_group_or_user string_or_placeholder role_options
> ```

I think you should be able to avoid having separate cases for no options, `role_options`, and `WITH role_options`. Check out how `opt_with_options` is defined.
___
*[pkg/sql/parser/sql.y, line 5128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pM2ps12PmqDqNYXu0:-M-pM2ps12PmqDqNYXu1:b-ne968s) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5128)):*
> ```Yacc
> USER { $$.val = false }
> | ROLE  { $$.val = true }
> | GROUP { $$.val = true }
> ```

I suspect we should keep the SKIP DOC comment here. It probably affects whether we generate documentation.
___
*[pkg/sql/roleoption/role_option.go, line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pNfX09ghx8gZT5grL:-M-pNfX09ghx8gZT5grM:buzc58k) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/roleoption/role_option.go#L150)):*
> ```Go
> }
> 
> func (ro RoleOption) toSQLValue() (string, error) {
> ```

unused?
___
*[pkg/sql/sem/tree/create.go, line 1172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pO3_J7e4MHhAU07eO:-M-pO3_J7e4MHhAU07eP:b-d8xh7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1172)):*
> ```Go
> 
> // OptionWithValue contains a RoleOption and it's value as an Expr.
> type OptionWithValue struct {
> ```

We should probably call this `RoleOptionWithValue` if it's role-specific.

I also noticed that there is already a `KVOption` type which serves a similar purpose for other commands (backup, restore, import). This includes validation via the `TypeAsStringOpts` function. Could you look into whether we could reuse that here rather than defining our own?
___
*[pkg/sql/sem/tree/create.go, line 1182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOBvND3oLVYHxuFSo:-M-pOBvND3oLVYHxuFSp:b-txax4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1182)):*
> ```Go
> // Format prints out the list in a buffer.
> // This keeps the existing order and uses " " as separator.
> func (rol OptionsWithValues) Format(ctx *FmtCtx) {
> ```

Nit: `rol` is kind of a weird name here. More standard to call it `o` or `ov`. Same goes for the other functions.
___
*[pkg/sql/sem/tree/create.go, line 1239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOSNT9JBjKFjgzAtM:-M-pOSNT9JBjKFjgzAtN:bhe5s9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1239)):*
> ```Go
> 
> // CreateUserOrRole represents a CREATE USER or CREATE ROLE statement.
> type CreateUserOrRole struct {
> ```

Like in `sql.y`, I'm wondering if it would be clearer to keep `CreateUser` and `CreateRole` separate here, even if it requires some slight code duplication.
___
*[pkg/sql/sem/tree/create.go, line 1269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOZzp48ih-yM4QvGH:-M-pOZzp48ih-yM4QvGI:by2kp93) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1269)):*
> ```Go
> 
> // AlterUserOrRoleOptions represents an ALTER ROLE or ALTER USER statement.
> type AlterUserOrRoleOptions struct {
> ```

How about just `AlterRole`?
___
*[pkg/sql/sem/tree/drop.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOd5VDxUidjjVmPvE:-M-pOd5VDxUidjjVmPvF:bg73gfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/drop.go#L143)):*
> ```Go
> 
> // DropUserOrRole represents a DROP USER statement
> type DropUserOrRole struct {
> ```

`DropRole`?
___
*[pkg/sql/alter_user.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pH6A2AVCcgdIPHgGY:-M-pH6A2AVCcgdIPHgGZ:bj98zgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43f506594e7561475cad4bf712c3a3329f0de35/pkg/sql/alter_user.go#L96)):*
> ```Go
> 		return err
> 	}
> 	if n.run.rowsAffected == 0 && !n.ifExists {
> ```

Do we still need the `ifExists` logic?
___
*[pkg/sql/alter_user_or_role.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGoiFEQIuXPSbsiG7:-M-pGoiFEQIuXPSbsiG8:b-33uw7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/alter_user_or_role.go#L26)):*
> ```Go
> 
> // alterUserOrRoleNode represents an ALTER ROLE ... [WITH] OPTION... statement.
> type alterUserOrRoleNode struct {
> ```

I would say let's just call this `alterRoleNode` now since ALTER USER will just be an alias for it.

**Review by RichardJCai — COMMENTED** [review]
For the first question, I think keeping them separate or merging them will depend on how we choose to handle some of the enterprise/non-enterprise stuff. If we choose to fully merge it, I think it would make sense to merge the syntax as they would function the same.

I'll look into using KVOption instead of OptionWithValue.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-:-M-pRnfEEnGdVdolI-bP:b-kvrjq6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai and @solongordon)

___
*[docs/generated/settings/settings.html, line  at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pBoepCPb4eLqynASB:-M-qDQbHB3qJnoCz7Q7u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/docs/generated/settings/settings.html#L)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Looks like this file went away for some reason?
</blockquote></details>

Done.
___
*[pkg/ccl/logictestccl/testdata/logic_test/role, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pC2Ub0zl1-xCfaRU0:-M-qCXQw4AsKYqCc2Je9:b-woq4nb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43f506594e7561475cad4bf712c3a3329f0de35/pkg/ccl/logictestccl/testdata/logic_test/role#L53)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Maybe we should still have a test for this if we intend for roles to be droppable with DROP USER?
</blockquote></details>

Yeah I was thinking about merging the tests for roles and users together once I get some more clarification on how to handle the separation between them. (For enterprise) 

If we choose to fully match pg, then a lot of the role/user separated stuff will be completely merged.

Added test back.
___
*[pkg/ccl/roleccl/role.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGK9_2AAVxWO34Yim:-M0t3NfI-pEnVMb-0o2j:bst7d19) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/ccl/roleccl/role.go#L116)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Should we just change the function name to `GetAllRoles`? And does it still need to return whether each is a user or a role?
</blockquote></details>

Updated map to behave as a set. All values of the map are true. Used to check for existence of usernames.
___
*[pkg/ccl/roleccl/role.go, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGHL84IAESvzIZ4yA:-M0t3N77AuDFpddCpqWs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/ccl/roleccl/role.go#L125)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

You can just update the comment.
</blockquote></details>

Done.
___
*[pkg/sql/alter_user.go, line 96 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pH6A2AVCcgdIPHgGY:-M0t5C0mCnnx42FF4niN:b-iwda8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43f506594e7561475cad4bf712c3a3329f0de35/pkg/sql/alter_user.go#L96)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Do we still need the `ifExists` logic?
</blockquote></details>

Added back
___
*[pkg/sql/database.go, line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pILDX0XLi-sb34OA6:-M0t5Hbb45aWhl_817zp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/database.go#L237)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Bad rename
</blockquote></details>

Done.
___
*[pkg/sql/parser/help_test.go, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJ-iZ8BKtivKaqiU3:-M0t5kxMArdM9w0eE8Ei:b750n0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/help_test.go#L69)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

This looks weird. If anything I would expect to show the ALTER ROLE help for both ALTER ROLE and ALTER USER. Though it might be nicer to keep them separate.
</blockquote></details>

Will keep them separate since we're still not allowing non-enterprise to use the CREATE ROLE syntax.
___
*[pkg/sql/parser/parse_test.go, line 1919 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJDih8_-5nKa0unGe:-M0t64CxAGOC6iWzpkwA:bn4iwpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/parse_test.go#L1919)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I still think it's fine if we just insert the WITH when it's not specified. *shrug*
</blockquote></details>

Updated to insert the WITH when it;'s not specified.
___
*[pkg/sql/parser/sql.y, line 346 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pLKwi2RikTWY9qok_:-M0tHd8t4QKcm_aVVPho:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L346)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I don't think you need to add this. You can just use the existing `bool()` method.
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 5072 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pJX-p2aB7OJOBi1DY:-M0tHxCyAkGZ4CLQSG2P:b-5ukkxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5072)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

It's unclear to me whether we should be merging these statements at the syntax level. CREATE USER is still slightly different from CREATE ROLE and we probably want to show different help text for the two. Do you think perhaps we should keep them separate here even though the underlying implementation is going to be the same?
</blockquote></details>

Yeah, going to split them mainly because to show different help text. Also because non-enterprise still won't have access to CREATE ROLE syntax
___
*[pkg/sql/parser/sql.y, line 5081 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pMgN04Oxm1QWpNKg-:-M0x02UdCMZ-Y_zAVsxd:b-ydv2mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5081)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I think you should be able to avoid having separate cases for no options, `role_options`, and `WITH role_options`. Check out how `opt_with_options` is defined.
</blockquote></details>

Good call, that really simplifies things.
___
*[pkg/sql/parser/sql.y, line 5128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pM2ps12PmqDqNYXu0:-M0x7vON2nUJTDlrFtCW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/parser/sql.y#L5128)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I suspect we should keep the SKIP DOC comment here. It probably affects whether we generate documentation.
</blockquote></details>

Done.
___
*[pkg/sql/roleoption/role_option.go, line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pNfX09ghx8gZT5grL:-M119eXO7Pa4dXvtRr8M:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/roleoption/role_option.go#L150)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Removed.
___
*[pkg/sql/sem/tree/create.go, line 1172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pO3_J7e4MHhAU07eO:-M0x818cDiBC37ecH4EG:b-s5jn2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1172)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

We should probably call this `RoleOptionWithValue` if it's role-specific.

I also noticed that there is already a `KVOption` type which serves a similar purpose for other commands (backup, restore, import). This includes validation via the `TypeAsStringOpts` function. Could you look into whether we could reuse that here rather than defining our own?
</blockquote></details>

KVOptions seems to work.
___
*[pkg/sql/sem/tree/create.go, line 1239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOSNT9JBjKFjgzAtM:-M0x8fH84NaWzsFyCR1U:bj1s8rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1239)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Like in `sql.y`, I'm wondering if it would be clearer to keep `CreateUser` and `CreateRole` separate here, even if it requires some slight code duplication.
</blockquote></details>

Yeah I think keeping them separate is a good idea as long as we keep the CreateRole version disabled for non-enterprise since it makes it easier to hide the CreateRole version behind the CCL license.

I went ahead and duplicated them.
___
*[pkg/sql/sem/tree/create.go, line 1269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOZzp48ih-yM4QvGH:-M0x86Qw3FJXbp6DUxN8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/create.go#L1269)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

How about just `AlterRole`?
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/drop.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pOd5VDxUidjjVmPvE:-M0x86o59chV-8b5-PPP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/sem/tree/drop.go#L143)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

`DropRole`?
</blockquote></details>

Done.
___
*[pkg/sql/alter_user_or_role.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pGoiFEQIuXPSbsiG7:-M0t3bDd48BxRS74Roaz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/alter_user_or_role.go#L26)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I would say let's just call this `alterRoleNode` now since ALTER USER will just be an alias for it.
</blockquote></details>

Done.

File: pkg/sql/logictest/testdata/logic_test/secondary_index_column_families [hunk 1]
```diff
@@ -1,107 +1,292 @@
-# Ensure updates that remove a k/v pair succeed.
```
Comment (rohany) [inline pkg/sql/logictest/testdata/logic_test/secondary_index_column_families]: Not sure why this change edits this file

**Review by jordanlewis — COMMENTED** [review]
This looks good to me besides the commit message, which needs to be updated since the functionality isn't added in this PR. I also have a couple of small nits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-:-M1fqMVJ3fQ_zkUUqSO_:b-ertfsv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @otan, @RichardJCai, and @solongordon)

___
*[pkg/ccl/roleccl/role.go, line 134 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fVecLEQMylstfjVfO:-M1fVecLEQMylstfjVfP:b-4qdevm) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/ccl/roleccl/role.go#L134)):*
> ```Go
> 		}
> 
> 		// Check that roles and roles exist.
> ```

This comment needs fixing
___
*[pkg/ccl/roleccl/role.go, line 286 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fW7ARBEuyckx2P4Ct:-M1fW7ARBEuyckx2P4Cu:b-fkfvxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/ccl/roleccl/role.go#L286)):*
> ```Go
> 		}
> 
> 		// Check that roles and roles exist.
> ```

comment needs updating
___
*[pkg/sql/alter_role.go, line 82 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fY1N7E6TbVjtXEEXW:-M1fY1N7E6TbVjtXEEXX:b-rsmyh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/sql/alter_role.go#L82)):*
> ```Go
> 
> // alterRoleRun is the run-time state of
> // AlterRoleNode for local execution.
> ```

nit: the struct is still called `alterRoleNode` no capitalization.
___
*[pkg/sql/sem/tree/create.go, line 1377 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fnHFEBUi9vVuW-ug4:-M1fnHFEBUi9vVuW-ug5:b-266s4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/sql/sem/tree/create.go#L1377)):*
> ```Go
> // Format implements the NodeFormatter interface.
> func (node *AlterUser) Format(ctx *FmtCtx) {
> 	// WIP: Need to fix this.
> ```

What needs to be fixed here?

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @otan, and @solongordon)

___
*[pkg/ccl/roleccl/role.go, line 134 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fVecLEQMylstfjVfO:-M1g0EgI-OdC_glCLcD6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/ccl/roleccl/role.go#L134)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This comment needs fixing
</blockquote></details>

Done.
___
*[pkg/ccl/roleccl/role.go, line 286 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fW7ARBEuyckx2P4Ct:-M1fusoz9wa7GZm_ZWh-:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/ccl/roleccl/role.go#L286)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

comment needs updating
</blockquote></details>

Done
___
*[pkg/sql/alter_role.go, line 82 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fY1N7E6TbVjtXEEXW:-M1fv1dQ9jY1r0rdqaVo:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/sql/alter_role.go#L82)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: the struct is still called `alterRoleNode` no capitalization.
</blockquote></details>

Fixed
___
*[pkg/sql/backfill.go, line 834 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M-pHJRiADdertfMFxrV:-M12OhWE3tvMO8eA7JaO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/401e5d747de44e82e7d2749da68c2ee68a42a6c9/pkg/sql/backfill.go#L834)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Bad rename? Many more examples of this in other files too. I won't comment on them all.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/create.go, line 1377 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-M1fnHFEBUi9vVuW-ug4:-M1fv4WSED6NoYR6bmVM:b9op7fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/87a25a2f9d4b1f2ef942dc814efaf308a2ddd18e/pkg/sql/sem/tree/create.go#L1377)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What needs to be fixed here?
</blockquote></details>

Oops, forgot to remove this comment.

**Review by jordanlewis — COMMENTED** [review]
I think the release note for this is still wrong.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-:-M1gACCe565UsulBl0o3:be3pkao)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @otan, and @solongordon)

**Review by jordanlewis — APPROVED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44968#-:-M1gZlmt3K-M0drSCFUg:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @otan, and @solongordon)

**Comment by RichardJCai** [comment]
Thanks for the review!
bors r=jordanlewis

**Pull request merged**